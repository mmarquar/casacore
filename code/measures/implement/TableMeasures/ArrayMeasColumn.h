//# ScalarMeasColumn.h: Definition of a Measure in a Table.
//# Copyright (C) 1997,1998
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#if !defined(AIPS_ARRAYMEASCOLUMN_H)
#define AIPS_ARRAYMEASCOLUMN_H

//# Includes
#include <aips/Measures/MeasRef.h>
#include <aips/Measures/MDirection.h>
#include <aips/Measures/MDoppler.h>
#include <aips/Measures/MEpoch.h>
#include <aips/Measures/MFrequency.h>
#include <aips/Measures/MPosition.h>
#include <aips/Measures/MRadialVelocity.h>
#include <aips/Measures/MBaseline.h>
#include <aips/Measures/Muvw.h>
#include <aips/Measures/MEarthMagnetic.h>
#include <aips/Tables/ArrayColumn.h>

//# Forward Declarations
class MEpoch;
class MVEpoch;
class MPosition;
class MVPosition;
class MDirection;
class MVDirection;
class MRadialVelocity;
class MVRadialVelocity;
class MDoppler;
class MVDoppler;
class MFrequency;
class MVFrequency;
class MBaseline;
class MVBaseline;
class Muvw;
class MVuvw;
class MEarthMagnetic;
class MVEarthMagnetic;
class String;
class Table;
template <class T> class ArrayColumn;
template <class T> class ScalarColumn;
template <class T> class ROScalarColumn;
template <class M> class TableMeasDesc;
template <class M, class MV> class ArrayMeasColumn;
template <class M, class MV> class ScalarMeasColumn;
template <class M, class MV> class ROScalarMeasColumn;

// <summary>
// Descinition of a Measure in a Table.
// </summary>

// <use visibility=export>

// <reviewed reviewer="" date="" tests="tTableMeasure.cc">
// </reviewed>

// <prerequisite>
//# Classes you should understand before using this one.
//   <li> <linkto class=MeasBase.h>MeasBase</linkto>
//   <li> <linkto class=ColumnDesc>ColumnDesc</linkto>
// </prerequisite>

// <synopsis>
// </synopsis>

// <example>
// <srcblock>
// </srcblock>
// </example>

// <motivation>
// Creating the required keyword for the definition of a Measure
// in a Table is somewhat complicated. This class assists in that
// process.
// </motivation>

//# <todo asof="$DATE:$">
//# A List of bugs, limitations, extensions or planned refinements.
//# </todo>

template<class M, class MV>
class ROArrayMeasColumn
{
public:
    // The default constructor creates a null object.  Useful for creating
    // arrays of ROArrayMeasColumn objects.  Attempting to use a null object
    // will produce a segmentation fault so care needs to be taken to
    // initialise the objects by using the attach() member before any attempt
    // is made to use the object.  A ROArrayQuantColumn object can be tested
    // for nullnes by using the isNull() member.
    ROArrayMeasColumn();

    // Create the ROArrayMeasColumn from the table and column Name.
    // Optionally specify an output reference. 
    ROArrayMeasColumn(const Table& tab, const String& columnName);

    // Copy constructor (copy semantics).
    ROArrayMeasColumn(const ROArrayMeasColumn<M, MV>& that);

    ~ROArrayMeasColumn();

    // Change the reference to another column.
    void reference(const ROArrayMeasColumn<M, MV>& that);

    // Attach a column to the object. 
    void attach(const Table& tab, const String& columnName); 
 
    // Get the measure in the default outref.
    void get(uInt rownr, Array<M>& meas, Bool resize = False) const;
    Array<M> operator()(uInt rownr) const;
          
    // Get the column's reference.
    const MeasRef<M>& getRef() const;
    
    // Test if the object is null.
    Bool isNull() const { return (itsDataCol == 0 ? True : False); }
    
    // Throw an exception if the object is null.
    void throwIfNull() const;
    
protected:
    // Resets itsMeasRef. Useful when the MeasRef varies from row to row.
    void setMeasRef(uInt rownr=0);

    // Measure reference could be constant or vary per row.
    Bool itsVarRefFlag;
    Bool itsVarOffsetFlag;
    
private:
    // Assignment makes no sense in a read only class.
    // Declaring this operator private makes it unusable.
    ROArrayMeasColumn& operator= (const ROArrayMeasColumn& that); 

    // Column which contains the Measure's actual data
    ROArrayColumn<Double>* itsDataCol;
    
    // Its MeasRef code column when references are variable.
    ROScalarColumn<Int>* itsRefIntCol;
    ROArrayColumn<Int>* itsArrRefIntCol;
    
    // Its MeasRef column when references are variable and stored as Strings.
    ROScalarColumn<String>* itsRefStrCol;
    ROArrayColumn<String>* itsArrRefStrCol;
    
    // Column containing its variable offsets.  Only applicable if the 
    // measure references have offsets and they are variable.
    ROScalarMeasColumn<M, MV>* itsOffsetCol;
    ROArrayMeasColumn<M, MV>* itsArrOffsetCol;
        
    // Its measure reference when the MeasRef is constant per row.
    MeasRef<M> itsMeasRef;
    
    // Deletes allocated memory etc. Called by ~tor and any member which needs
    // to reallocate data.
    void cleanUp();
};

typedef ROArrayMeasColumn<MEpoch, MVEpoch> ROMEpochArrCol;
typedef ROArrayMeasColumn<MDirection, MVDirection> ROMDirectionArrCol;
typedef ROArrayMeasColumn<MPosition, MVPosition> ROMPositionArrCol;
typedef ROArrayMeasColumn<MRadialVelocity, MVRadialVelocity> 
    ROMRadialVelocityArrCol;
typedef ROArrayMeasColumn<MDoppler, MVDoppler> ROMDopplerArrCol;
typedef ROArrayMeasColumn<MFrequency, MVFrequency> ROMFrequecnyCol;
typedef ROArrayMeasColumn<MBaseline, MVBaseline> ROMBaselineArrCol;
typedef ROArrayMeasColumn<Muvw, MVuvw> ROMuvwArrCol;
typedef ROArrayMeasColumn<MEarthMagnetic, MVEarthMagnetic> ROMEarthMagneticArrCol;


template<class M, class MV>
class ArrayMeasColumn : public ROArrayMeasColumn<M, MV>
{
public:
    // The default constructor creates a null object.  Useful for creating
    // arrays of ROArrayMeasColumn objects.  Attempting to use a null object
    // will produce a segmentation fault so care needs to be taken to
    // initialise the objects by using the attach() member before any attempt
    // is made to use the object.  A ROArrayQuantColumn object can be tested
    // for nullnes by using the isNull() member.
    ArrayMeasColumn();

    // Create the ROArrayMeasColumn from the table and column Name.
    // Optionally specify an output reference. 
    ArrayMeasColumn(const Table& tab, const String& columnName);

    // Copy constructor (copy semantics).
    ArrayMeasColumn(const ArrayMeasColumn<M, MV>& that);

    ~ArrayMeasColumn();

    // Change the reference to another column.
    void reference(const ArrayMeasColumn<M, MV>& that);

    // Attach a column to the object. 
    void attach(const Table& tab, const String& columnName); 
 
    // Add a measure to the table
    void put(uInt rownr, const Array<M>&);    
private:
    // See class <linkto class="ArrayColumn">ArrayColumn</linkto> for an
    // explanation as to why this operation is disallowed.  Use the reference
    // function instead.
    // Declaring this operator private makes it unusable.
    ArrayMeasColumn& operator= (const ArrayMeasColumn& that); 

    // Column which contains the Measure's actual data
    ArrayColumn<Double>* itsDataCol;
    
    // Its MeasRef column when references are variable and stored as RefCodes.
    ScalarColumn<Int>* itsRefIntCol;
    ArrayColumn<Int>* itsArrRefIntCol;
    
    // Its MeasRef column when references are variable and stored as Strings.
    ScalarColumn<String>* itsRefStrCol;
    ArrayColumn<String>* itsArrRefStrCol;
    
    // Column containing its variable offsets.  Only applicable if the 
    // measure references have offsets and they are variable.
    ScalarMeasColumn<M, MV>* itsOffsetCol;
    ArrayMeasColumn<M, MV>* itsArrOffsetCol;

    // Its measure reference when the MeasRef is constant per row.
    MeasRef<M> itsMeasRef;
        
    // Deletes allocated memory etc. Called by ~tor and any member which needs
    // to reallocate data.
    void cleanUp();
};

typedef ArrayMeasColumn<MEpoch, MVEpoch> MEpochArrCol;
typedef ArrayMeasColumn<MDirection, MVDirection> MDirectionArrCol;
typedef ArrayMeasColumn<MPosition, MVPosition> MPositionArrCol;
typedef ArrayMeasColumn<MRadialVelocity, MVRadialVelocity> 
    MRadialVelocityArrCol;
typedef ArrayMeasColumn<MDoppler, MVDoppler> MDopplerArrCol;
typedef ArrayMeasColumn<MFrequency, MVFrequency> MFrequecnyCol;
typedef ArrayMeasColumn<MBaseline, MVBaseline> MBaselineArrCol;
typedef ArrayMeasColumn<Muvw, MVuvw> MuvwArrCol;
typedef ArrayMeasColumn<MEarthMagnetic, MVEarthMagnetic> MEarthMagneticArrCol;

#endif
