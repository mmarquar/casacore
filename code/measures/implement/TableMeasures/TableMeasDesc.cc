//# TableMeasDef.cc: Definition of a Measure in a Table.
//# Copyright (C) 1997,1999
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <trial/TableMeasures/TableMeasDesc.h>
#include <aips/Measures/MeasRef.h>
#include <aips/Quanta/Unit.h>
#include <aips/Tables/TableDesc.h>
#include <aips/Arrays/Vector.h>
#include <aips/Exceptions/Error.h>


template<class M>
TableMeasDesc<M>::TableMeasDesc (const TableMeasValueDesc& value)
: TableMeasDescBase(value, TableMeasRefDesc(M::DEFAULT))
{
  M meas;
  Vector<Quantum<Double> > val;
  val = meas.getValue().getRecordValue();
  Vector<Unit> u;
  setMeasUnits (meas, val, u);
}

template<class M>
TableMeasDesc<M>::TableMeasDesc (const TableMeasValueDesc& value,
				 const Vector<Unit>& u)
: TableMeasDescBase(value, TableMeasRefDesc(M::DEFAULT))
{ 
  M meas;
  Vector<Quantum<Double> > val;
  val = meas.getValue().getRecordValue();
  setMeasUnits (meas, val, u);
}

template<class M>
TableMeasDesc<M>::TableMeasDesc (const TableMeasValueDesc& value,
				 const TableMeasRefDesc& ref)
: TableMeasDescBase(value, ref)
{
  M meas;
  Vector<Quantum<Double> > val;
  val = meas.getValue().getRecordValue();
  Vector<Unit> u;
  setMeasUnits (meas, val, u);
}

template<class M>
TableMeasDesc<M>::TableMeasDesc (const TableMeasValueDesc& value,
				 const TableMeasRefDesc& ref,
				 const Vector<Unit>& u)
: TableMeasDescBase(value, ref)
{
  M meas;
  Vector<Quantum<Double> > val;
  val = meas.getValue().getRecordValue();
  setMeasUnits (meas, val, u);
}

template<class M>
TableMeasDescBase* TableMeasDesc<M>::clone() const
{
  return new TableMeasDesc<M>(*this);
}

template<class M>
TableMeasDesc<M>::TableMeasDesc (const TableMeasDesc<M>& that)
: TableMeasDescBase(that)
{}

template<class M>
TableMeasDesc<M>::~TableMeasDesc()
{}

template<class M>
TableMeasDesc<M>& TableMeasDesc<M>::operator= (const TableMeasDesc<M>& that)
{
  TableMeasDescBase::operator= (that);
  return *this;
}
