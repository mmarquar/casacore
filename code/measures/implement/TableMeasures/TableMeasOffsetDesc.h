//# TableMeasOffseDesc.h: Definition of aOffset measure in a Table.
//# Copyright (C) 1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#if !defined(AIPS_TABLEMEASOFFSETDESC_H)
#define AIPS_TABLEMEASOFFSETDESC_H

//# Includes

//# Forward Declarations
class Measure;
class String;
class Table;
class TableMeasDescBase;
class TableRecord;

// <summary>
// Definition of a Measure Offset in a Table.
// </summary>

// <use visibility=export>

// <reviewed reviewer="" date="" tests="tTableMeasure.cc">
// </reviewed>

// <prerequisite>
//# Classes you should understand before using this one.
//   <li> <linkto module=Measures>Measures</linkto>
//   <li> <linkto module=Tables>Tables</linkto>
//   <li> TableMeasDesc
// </prerequisite>

// <synopsis>
// The measure offset can be static for an entire column or it can vary per
// row.
// </synopsis>

// <example>
// See class <linkto class="TableMeasDesc">TableMeasDesc</linkto>.
// </example>

// <motivation>
// Creating the required keyword for the definition of a Measure
// in a Table is somewhat complicated. This class assists in that
// process.
// </motivation>

//# <todo asof="$DATE:$">
//# A List of bugs, limitations, extensions or planned refinements.
//# </todo>

class TableMeasOffsetDesc
{
public:
    // Constructor which defines a constant (non-variable) offset.
    TableMeasOffsetDesc(const Measure& offset);

    // Constructor for defining a variable offset.  
    TableMeasOffsetDesc(const TableMeasDescBase& offsetColumn);

    // Copy constructor.
    TableMeasOffsetDesc(const TableMeasOffsetDesc& that);

    ~TableMeasOffsetDesc();

    // Assignment operator.
    TableMeasOffsetDesc& operator=(const TableMeasOffsetDesc& that);
    
    // Reconstructs the TableMeasOffsetDesc from the measInfo TableRecord.
    static TableMeasOffsetDesc* reconstruct(const TableRecord& measInfo,
				    	    const String& prefix,
		    	    	            const Table& tab);
    
    // Get the (non-variable) measure offset for this column.  If it doesn't
    // exist (the offset is variable) and exception is thrown.
    const Measure& getOffset() const;
    
    // Returns True is the offset varies per row.
    Bool isVariable() const { return (itsTMDesc != 0 ? True : False); }
    
    // Gets the name of the column which stores the variable offset. "" is
    // returned if the offset is not variable.
    const String& columnName() const;
    
    // Write the information into the record.
    void write(TableDesc& td, TableRecord& measInfo, const String& prefix);

private:
    // Constructor which uses the measInfo TableRecord.
    TableMeasOffsetDesc(const TableRecord& measInfo, const String& prefix,
		      	const Table&);

    TableMeasDescBase* itsTMDesc;   //# Stores variable offset if applicable
    Measure* itsMeasure;    	    //# The offset if non-variable.
    String itsVarColName;   	    //# "" if offset non-variable.
};

#endif
