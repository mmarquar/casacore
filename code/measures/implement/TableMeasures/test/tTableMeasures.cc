//# tTableMeasures.cc: test program for the TableMeasures class.
//# Copyright (C) 1994,1995,1996,1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <iostream.h>
#include <aips/aips.h>
#include <aips/Exceptions.h>
#include <aips/Arrays/Array.h>
#include <aips/Arrays/Vector.h>
#include <aips/Measures/MeasConvert.h>
#include <aips/Measures/MeasFrame.h>
#include <aips/Measures/MEpoch.h>
#include <aips/Measures/MVEpoch.h>
#include <aips/Measures/MVTime.h>
#include <aips/Measures/MDirection.h>
#include <aips/Measures/MFrequency.h>
#include <aips/Measures/MPosition.h>
#include <aips/Measures/MRadialVelocity.h>
#include <aips/Measures/MeasData.h>
#include <aips/Measures/MeasRef.h>
#include <aips/Tables/Table.h>
#include <aips/Tables/TableDesc.h>
#include <aips/Tables/ArrayColumn.h>
#include <aips/Tables/ColumnDesc.h>
#include <aips/Tables/ArrColDesc.h>
#include <aips/Tables/SetupNewTab.h>
#include <aips/Tables/ScaColDesc.h>
#include <aips/Tables/ScalarColumn.h>
#include <aips/Tables/TableRecord.h>
#ifdef COMMENT
#include <trial/TableMeasures/ArrayMeasColumn.h>
#endif
#include <trial/TableMeasures/ScalarMeasColumn.h>
#include <trial/TableMeasures/TableMeasValueDesc.h>
#include <trial/TableMeasures/TableMeasOffsetDesc.h>
#include <trial/TableMeasures/TableMeasRefDesc.h>
#include <trial/TableMeasures/TableMeasDesc.h>

int main(void)
{
  try {
    cout << "Test of TableMeasures classes.\n";

    // Need a table to work with.
    TableDesc td("tTableMeasure_desc", "1", TableDesc::New);
    td.comment() = "A test of TableMeasures class.";
    
    // The underlying column for storing the Measure values is an Array<Double>
    // Create an underlying column for each Measure column.
    ArrayColumnDesc<Double> cdTime("Time1", "An MEpoch column");
    ArrayColumnDesc<Double> cdTOffset("TimeOffset", "Offsetcolumn for Time1");
    ArrayColumnDesc<Double> cdVarOffset("TimeVarOffset", "Variable Offset col");
    ArrayColumnDesc<Double> cdMPos("MPosColumn", "Simple mposition column");
    ScalarColumnDesc<Int> cdTimeRef("TimeRef", "Reference column for Time1");
    
#ifdef ARRAYMEAS
    ArrayColumnDesc<Double> cdTimeArr("Time1Arr", "An MEpoch array column");
#endif

    td.addColumn(cdTime);
    td.addColumn(cdTOffset);
    td.addColumn(cdVarOffset);
    td.addColumn(cdMPos);
    td.addColumn(cdTimeRef);
    
#ifdef ARRAYMEAS
    td.addColumn(cdTimeArr);
#endif
    
    {
    	// Another MEpoch column descriptor this one specifies a fixed offset 
	// with reference MEpoch::LAST.
	MEpoch obsTime(MVEpoch(MVTime(1996, 5, 17, (8+18./60.)/24.)),
		MEpoch::UTC);
	TableMeasOffsetDesc tmodObsTime(obsTime);
	TableMeasRefDesc tmrdObs(MEpoch::LAST, tmodObsTime);
	TableMeasValueDesc tmvdObs(td, "TimeOffset");    
	TableMeasDesc<MEpoch> tmdObs(tmvdObs, tmrdObs);
	
	// test assignment and copy constructor
	TableMeasDesc<MEpoch> tmdObs1 = tmdObs;
	TableMeasDesc<MEpoch> tmdObs2(tmdObs1);
	tmdObs2.write(td);
    }
    
    {
    	// Set up a MEpoch column with support for variable references and
	// offset
	
	// A variable offset column is a Measure column, so a TableMeasDesc
	// is needed.
	TableMeasValueDesc tmvdObs(td, "TimeVarOffset");    
	TableMeasDesc<MEpoch> tmMOS(tmvdObs);
	
        // This sets up a second MEpoch column.  Variable references are 
	// specified by supplying the name of a reference column "TimeRef" to
	// use.  tmMOS is used as a variable measure offset column.
	TableMeasRefDesc tmrd("TimeRef", tmMOS);
	TableMeasValueDesc tmvd(td, "Time1");    
	TableMeasDesc<MEpoch> tmdMEpoch(tmvd, tmrd);
	tmdMEpoch.write(td);
    }
    
    {
	// Simplest useful TableMeasDesc declaration that can be done.  
	// Default MPosition::Ref will be used for the reference.
	TableMeasValueDesc tmvdMPos(td, "MPosColumn");    
	TableMeasDesc<MPosition> tmdMPos(tmvdMPos);
	tmdMPos.write(td);
    }
    
    
#ifdef ARRAYMEAS
    {
	// An array MEpoch column descriptor.  The TableMeasDesc for an Array
	// measure column is identical to the Scalar measure column.
	
	MEpoch mjdToday(MVEpoch(51234));
	TableMeasOffsetDesc tmodToday(mjdToday);
	TableMeasRefDesc tmrdLast(MEpoch::LAST, tmodToday);
	TableMeasValueDesc tmvdLast(td, "Time1Arr");
	// create a tmp and test if copy constructor and assignment work
	TableMeasDesc<MEpoch> tmp(tmvdLast, tmrdLast);
	TableMeasDesc<MEpoch> tmp2 = tmp;
	TableMeasDesc<MEpoch> tmdArrLast(tmp2);	
	
	tmdArrLast.write(td);
    }    
#endif    
    // create the table
    SetupNewTable newtab("TestTableMeasures", td, Table::New);
    const uInt tabRows = 5;
    Table tab(newtab, tabRows);
    
    {
	cout << "Create MEpochScaCol from column TimeOffset...\n";
	cout << "A column of MEpochs where the reference and offset are ";
	cout << "non-variable.\n";
	// create first a null object and attach it and copy it etc.
	// show that these operation work.
	MEpochScaCol tmpCol;
	if (tmpCol.isNull()) {
	    tmpCol.attach(tab, "TimeOffset");
	}
	tmpCol.throwIfNull();
	cout << "Null MEpochScaCol successfully attached\n";
	// no assignment operator but there is a copy constructor
	MEpochScaCol timeCol(tmpCol);
	
	// print some details things about the column
	if (timeCol.isRefVariable()) {
	    cout << "The column has variable references.\n";
	} else {
	    cout << "The MeasRef for the column is: " << timeCol.getMeasRef() 
		<< endl;
	}
	
	cout << "Adding a few MEpochs to column TimeOffset...\n";
	MEpoch tm(MVEpoch(1234.));
	uInt i;
	for (i=0; i<tabRows; i++) {
	    tm.set(MVEpoch(1234 + (i/10.0)));
	    cout << "adding: " << tm << endl;
	    timeCol.put(i, tm);
	}

	cout << "Reading the MEpochs back from TimeOffset...\n";
	// Create read only measure column
	ROMEpochScaCol timeColRead(tab, "TimeOffset");
	for (i=0; i<tabRows; i++) {
	    if (timeColRead.isDefined(i)) {
		timeColRead.get(tm, i);
	    	cout << "retrieve: " << tm << endl;
		cout << "  " << tm.getRef() << endl;
    	    } else {
		cout << "Error: row " << i << " doesn't contain nay data!\n";
	    }
	}
	
    }
    
    {
	// A column of MPositions
	MPositionScaCol mposCol(tab, "MPosColumn");
	if (mposCol.isRefVariable()) {
	    cout << "Error: reference is variable!\n";
	}
	
	cout << "Filling the MPosition column MPosColumn\n";
	MPosition mpos;
	for (uInt i=0; i<tabRows; i++) {
	    mpos.set(MVPosition(Quantity(25 + i, "m"),
		    	    	Quantity(20, "deg"),
		    	    	Quantity(53, "deg")));
	    cout << "put: " << mpos << endl;	    
	    mposCol.put(i, mpos);
	}
    }
    
    {
	// A column of MPositions
	ROMPositionScaCol mposCol(tab, "MPosColumn");
	if (mposCol.isRefVariable()) {
	    cout << "Error: reference is variable!\n";
	}
	MPosition mpos;
	cout << "Reading from MPosition column MPosColumn\n";
	for (uInt i=0; i<tabRows; i++) {
	    cout << "retrieve: " << mposCol(i) << endl;
	}
    }
    
    {
	cout << "Test of column TimeVarOffset...\n";
	cout << "A column of MEpoch where the reference and offset components";
	cout << " are variable.\n";
	MEpochScaCol meCol(tab, "Time1");
	
	MEpoch offset;
	MEpoch me;
	offset.set(MVEpoch(1234));
	me.set(MVEpoch(51234.1), MEpoch::Ref(MEpoch::LAST, offset));
	meCol.put(0, me);
	offset.set(MVEpoch(1234.1));
	me.set(MVEpoch(51234.2), MEpoch::Ref(MEpoch::UTC, offset));
	meCol.put(1, me);
	offset.set(MVEpoch(1234.2));
	me.set(MVEpoch(51234.3), MEpoch::Ref(MEpoch::TAI, offset));
	meCol.put(2, me);
	offset.set(MVEpoch(1234.3));
	me.set(MVEpoch(51234.4), MEpoch::Ref(MEpoch::UTC, offset));
	meCol.put(3, me);	
	offset.set(MVEpoch(1234.4));
	me.set(MVEpoch(51234.5), MEpoch::Ref(MEpoch::LAST, offset));
	meCol.put(4, me);
	
	for (uInt i=0; i<tabRows; i++) {
	    meCol.get(me, i);
	    cout << "retrieve: " << me << endl;
	    cout << "  " << me.getRef() << endl;
	}
	
    }
	

#ifdef ARRAYMEAS
    {
	cout << "Creating an MEpoch Array Column\n";
	MEpochArrCol arrayCol(tab, "Time1Arr");

	MEpoch last(Quantity(13.45, "h"), MEpoch::Ref(MEpoch::TAI));
	Vector<MEpoch> ev(10);
	for (uInt i=0; i<10; i++) {
	    last.set(Quantity(13.45 + i, "h"));
	    ev(i) = last;
	}
	
	cout << "Adding a vector to the column at row 0.\n";
	arrayCol.put(0, ev);
	Vector<MEpoch> ew;
	arrayCol.get(ew, 0, True);
	
	for (i=0; i<10; i++) {
	    cout << "ev: " << ev(i) << " " << ev(i).getRef() << endl;
	    cout << "ew: " << ew(i) << " " << ew(i).getRef() << endl;
	}
//	cout << arrayCol(0) << endl;
    }
#endif
    
    cout << "Bye.\n";  
  } catch (AipsError x) {
    cout << x.getMesg() << endl;
  } end_try;
    exit(0);
}


