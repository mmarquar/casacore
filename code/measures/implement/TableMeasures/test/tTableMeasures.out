Test of TableMeasures classes.
Create MEpochScaCol from column Time1...
Null MEpochScaCol successfully attached
The MeasRef for the column is: Reference for an Epoch with Type: LAST, Offset: Epoch: 50220::08:17:60.0000
Epoch: 50220::10:47:60.0000
Epoch: 50220::10:47:60.0000
Epoch: 50220::10:47:60.0000
Epoch: 50220::10:47:60.0000
Epoch: 50220::10:47:60.0000
Creating an MEpoch Array Column
Adding a vector to the column at row 0.
elements: 10
ndim(): 1
new ms shape: [10, 2]
array.nelements(): 20
ev: Epoch: 0::13:26:60.0000 Reference for an Epoch with Type: TAI
ew: Epoch: 0::13:26:60.0000 Reference for an Epoch with Type: LAST, Offset: Epoch: 51234::00:00:00.0000
ev: Epoch: 0::14:26:60.0000 Reference for an Epoch with Type: TAI
ew: Epoch: 0::14:26:60.0000 Reference for an Epoch with Type: LAST, Offset: Epoch: 51234::00:00:00.0000
ev: Epoch: 0::15:27:00.0000 Reference for an Epoch with Type: TAI
ew: Epoch: 0::15:27:00.0000 Reference for an Epoch with Type: LAST, Offset: Epoch: 51234::00:00:00.0000
ev: Epoch: 0::16:26:60.0000 Reference for an Epoch with Type: TAI
ew: Epoch: 0::16:26:60.0000 Reference for an Epoch with Type: LAST, Offset: Epoch: 51234::00:00:00.0000
ev: Epoch: 0::17:26:60.0000 Reference for an Epoch with Type: TAI
ew: Epoch: 0::17:26:60.0000 Reference for an Epoch with Type: LAST, Offset: Epoch: 51234::00:00:00.0000
ev: Epoch: 0::18:27:00.0000 Reference for an Epoch with Type: TAI
ew: Epoch: 0::18:27:00.0000 Reference for an Epoch with Type: LAST, Offset: Epoch: 51234::00:00:00.0000
ev: Epoch: 0::19:26:60.0000 Reference for an Epoch with Type: TAI
ew: Epoch: 0::19:26:60.0000 Reference for an Epoch with Type: LAST, Offset: Epoch: 51234::00:00:00.0000
ev: Epoch: 0::20:26:60.0000 Reference for an Epoch with Type: TAI
ew: Epoch: 0::20:26:60.0000 Reference for an Epoch with Type: LAST, Offset: Epoch: 51234::00:00:00.0000
ev: Epoch: 0::21:27:00.0000 Reference for an Epoch with Type: TAI
ew: Epoch: 0::21:27:00.0000 Reference for an Epoch with Type: LAST, Offset: Epoch: 51234::00:00:00.0000
ev: Epoch: 0::22:26:60.0000 Reference for an Epoch with Type: TAI
ew: Epoch: 0::22:26:60.0000 Reference for an Epoch with Type: LAST, Offset: Epoch: 51234::00:00:00.0000
Bye.
